src/full_coverage_path_planner.cpp

```diff
@@ -32,218 +32,217 @@
+    if (goalpoints.size() > 1)
+    {
+      for (it = goalpoints.begin(); it != goalpoints.end(); ++it)
       {
-        new_goal.header.frame_id = "map";
-        new_goal.pose.position.x = (it->x) * tile_size_ + grid_origin_.x + tile_size_ * 0.5;
-        new_goal.pose.position.y = (it->y) * tile_size_ + grid_origin_.y + tile_size_ * 0.5;
-        // Calculate desired orientation to be in line with movement direction
-        switch (move_dir_now)
+        it_next = it;
+        it_next++;
+        it_prev = it;
+        it_prev--;
+
+        // Check for the direction of movement
+        if (it == goalpoints.begin())
         {
-        case eDirNone:
-          // Keep orientation
-          break;
-        case eDirRight:
-          orientation = 0;
-          break;
-        case eDirUp:
-          orientation = M_PI / 2;
-          break;
-        case eDirLeft:
-          orientation = M_PI;
-          break;
-        case eDirDown:
-          orientation = M_PI * 1.5;
-          break;
+          dx_now = it_next->x - it->x;
+          dy_now = it_next->y - it->y;
         }
-        new_goal.pose.orientation = createQuaternionMsgFromYaw(orientation);
-        if (it != goalpoints.begin())
+        else
         {
-          previous_goal_.pose.orientation = new_goal.pose.orientation;
-          // republish previous goal but with new orientation to indicate change of direction
-          // useful when the plan is strictly followed with base_link
-          plan.push_back(previous_goal_);
+          dx_now = it->x - it_prev->x;
+          dy_now = it->y - it_prev->y;
+          dx_next = it_next->x - it->x;
+          dy_next = it_next->y - it->y;
+        }
+
+        // Calculate direction enum: dx + dy*2 will give a unique number for each of the four possible directions because
+        // of their signs:
+        //  1 +  0*2 =  1
+        //  0 +  1*2 =  2
+        // -1 +  0*2 = -1
+        //  0 + -1*2 = -2
+        move_dir_now = dx_now + dy_now * 2;
+        move_dir_next = dx_next + dy_next * 2;
+
+        // Check if this points needs to be published (i.e. a change of direction or first or last point in list)
+        do_publish = move_dir_next != move_dir_now || it == goalpoints.begin() ||
+                     (it != goalpoints.end() && it == --goalpoints.end());
+
+        // Add to vector if required
+        if (do_publish)
+        {
+          new_goal.header.frame_id = "map";
+          new_goal.pose.position.x = (it->x) * tile_size_ + grid_origin_.x + tile_size_ * 0.5;
+          new_goal.pose.position.y = (it->y) * tile_size_ + grid_origin_.y + tile_size_ * 0.5;
+          // Calculate desired orientation to be in line with movement direction
+          switch (move_dir_now)
+          {
+          case eDirNone:
+            // Keep orientation
+            break;
+          case eDirRight:
```

Comment:
Use a c++ class enum. This then becomes `dir::right` which is just as readible but the names are scoped.